main.py

```
@@ -1,69 +1,307 @@
+        
+        for attempt in range(self.max_retries):
+            try:
+                self.throttler.throttle()
+                response = self.session.get(url, timeout=self.timeout)
+                response.raise_for_status()
+                logger.info(f"Successfully retrieved the website content (status code: {response.status_code}).")
+                return response.text
+            except requests.exceptions.HTTPError as http_err:
+                logger.warning(f"HTTP error on attempt {attempt+1}/{self.max_retries}: {http_err}")
+                if attempt == self.max_retries - 1:
+                    logger.error(f"Failed to retrieve {url} after {self.max_retries} attempts.")
```

Ruff (0.8.2)

Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)